Make sure user can build in the project before uploading SRPM

FrostyX opened the pull request:
Fix #134

Tested for both non-existing project and someone else's project that I don't have permissions to.

We can easily drop the `self.client.base_proxy.auth_check()` from CLI because there is `@requires_api_auth` decorator on the function.

praiskup commented on frontend/coprs_frontend/coprs/views/apiv3_ns/apiv3_permissions.py:
```diff
@@ -11,6 +11,20 @@
 from . import GET, PUT, editable_copr, get_copr
 
 
+@apiv3_ns.route("/project/permissions/can_build_in/<ownername>/<projectname>")
```
Perhaps a small thing: When I see `can_build_in`, I don't know who.  From the method name I'd expect a username as an argument.  Should we have a better name?

FrostyX replied:
Interesting perspective. I thought about it as "Can I (as a logged user) build in <ownername>/<projectname>". But you are right, why limitting this to _myself_ only. 

Maybe we can remove the `@api_login_required` decorator making this route available for everyone and instead, adding a `username` parameter?

FrostyX replied:
What I want to change is the behavior of `ProjectProxy.can_build_in` method. Either, I want to rename it to `assert_can_build_in` changing it to return a boolean.

But I will take a look what should we return in the API call as well.

praiskup commented on frontend/coprs_frontend/coprs/views/apiv3_ns/apiv3_permissions.py:
```diff
@@ -11,6 +11,20 @@
+
+    # I don't really know what to return. Perhaps we can start with an empty
+    # dict and add information as we need it? For now, it's important that it
+    # returns 200 status code.
```
The call has "assert"-like, which is a bit uncommon in APIs.  We also return BadRequest (error 400).
I'm curious if there's some "guidance" in some API best-practices documentation...?

FrostyX commented:
I addressed the assert-like call

> Perhaps a small thing: When I see can_build_in, I don't know who. From the method name I'd expect a username as an argument.

I added another parameter for specifying who is asking. I kind of like both more and less at the same time, so it is in a separate commit. Should I use it or delete it?

> Should we have a better name?

Fine with me but I couldn't think of any better

praiskup commented on cli/copr_cli/main.py:
```diff
@@ -308,12 +308,15 @@ def action_build(self, args):
 
-        # Before we start uploading potentially large source RPM file, make sure
-        # that the user actually has a valid credentials.
-        self.client.base_proxy.auth_check()
```
This removal has a small potential to cause trouble.  It actually did a real `/auth-check` call before - revealing that session expired (e.g. wiped-out redis DB because frontend restarted).

Keeping this has an obvious consequence - additional API call for each build request.

FrostyX replied:
Good catch. 
There are two commits in this PR. This issue can be easily solved by removing the second commit. We can do it that way or any other way you prefer.

praiskup commented on cli/copr_cli/main.py:
```diff
@@ -308,12 +308,15 @@ def action_build(self, args):
+        if not self.client.project_proxy.can_build_in(username, projectname):
+            msg = ("User '{0}' is not allowed to build in '{1}/{2}'"
+                   .format(self.username, username, projectname))
+            raise CoprRequestException(msg)
```
Awesome, thank you!  KISS way of catching a user's mistake.

praiskup commented on frontend/coprs_frontend/coprs/views/apiv3_ns/apiv3_permissions.py:
```diff
@@ -11,6 +11,14 @@
 
+@apiv3_ns.route("/project/permissions/can_build_in/<ownername>/<projectname>")
+@api_login_required
+def can_build_in(ownername, projectname):
```
missing doc string

FrostyX replied:
Something weird happened and you commented on an outdated code. The docstring is there.

praiskup commented on cli/tests/test_cli.py:
```diff
@@ -577,11 +577,13 @@ def test_create_multilib_project(config_from_file, project_proxy_add, capsys):
 
-@mock.patch('copr.v3.proxies.BaseProxy.auth_check', return_value=Munch(name="test"))
+@mock.patch('copr.v3.proxies.project.ProjectProxy.can_build_in',
+            return_value=Munch(name="test"))
```
`Munch(name="test")` as `can_build_in` raises questions :)

FrostyX commented:
Updated according to what we discussed on the meeting.

praiskup commented on frontend/coprs_frontend/coprs/views/apiv3_ns/apiv3_permissions.py:
```diff
@@ -5,12 +5,29 @@
+        "projectname": projectname,
+        "can_build_in": user.can_build_in(copr),
+    }
+    return flask.jsonify(result)
```
Does not sqlalchemy sanitize this for us properly?
